companion/src/shared/timeredit.cpp

```diff
@@ -0,0 +1,281 @@
+  m_hourDigits(2),
+  m_validator(new QRegularExpressionValidator(this))
+{
+  setSizePolicy(QSizePolicy::Minimum, QSizePolicy::Fixed);
+  setupFormat();
+  setToolTip(tr("Use mouse scroll or UP/DOWN arrow keys to change time in small steps.\nCTRL + scroll or PAGE UP/DOWN keys to change time in larger steps."));
+  connect(this, &TimerEdit::textEdited, this, &TimerEdit::textEditedHandler);
+}
+
+int TimerEdit::timeInSeconds() const
+{
+  int ret = 0;
+
+  ret += timepart(text(), TimeParts::Hours) * 3600;
+  ret += timepart(text(), TimeParts::Minutes) * 60;
+  ret += timepart(text(), TimeParts::Seconds);
+  ret *= timepart(text(), TimeParts::Polarity);
+
+  //qDebug() << ret;
+  return ret;
+}
+
+int TimerEdit::timepart(const QString & input, TimerEdit::TimeParts part) const
+{
+  int ret = 0;
+  QRegularExpressionMatch match = m_validator->regularExpression().match(input, 0, QRegularExpression::NormalMatch);
+
+  if (match.hasMatch()) {
+    switch (part) {
+      case Hours:
+        if (!match.captured("hrs").isEmpty())
+          ret = match.captured("hrs").toInt();
+        break;
+
+      case Minutes:
+        if (!match.captured("mins").isEmpty())
+          ret = match.captured("mins").toInt();
+        break;
+
+      case Seconds:
+        if (!match.captured("secs").isEmpty())
+          ret = match.captured("secs").toInt();
+        break;
+
+      case Polarity:
+        if (!match.captured("pol").isEmpty() && !match.captured("pol").compare("-"))
+          ret = -1;
+        else
+          ret = 1;
+        break;
+
+      default:
+        break;
+    }
+  }
+  //qDebug() << input << match << ret;
+
+  return ret;
+}
+
+
+//// Slots ////
+
+void TimerEdit::setTime(int seconds)
+{
+  if (seconds > m_maxTime)
+    seconds = m_maxTime;
+  else if (seconds < m_minTime)
+    seconds = m_minTime;
+
+  div_t qrM = div(abs(seconds), 60);
+  div_t qrH = div(qrM.quot, 60);
+
+  QString val = QString("%1").arg(qrH.rem, 2, 10, QChar('0'));
+
+  if (m_hourDigits)
+    val.prepend(QString("%1:").arg(qrH.quot, m_hourDigits, 10, QChar('0')));
+
+  if (m_showSeconds)
+    val.append(QString(":%1").arg(qrM.rem, 2, 10, QChar('0')));
+
+  if (m_minTime < 0) {
+    if (seconds < 0)
+      val.prepend("-");
+    else
+      val.prepend(" ");
+  }
+  //qDebug() << seconds << qrH.quot << qrH.rem << qrM.quot << qrM.rem << val << text();
+
+  if (text().compare(val)) {
+    setText(val);
+  }
+}
+
+void TimerEdit::incrDecr(int seconds)
+{
+  if (!seconds)
+    return;
+
+  const int cs = timeInSeconds();
+  int s = seconds + cs;
+  if (s < m_minTime)
+    s = m_minTime;
+  if (s > m_maxTime)
+    s = m_maxTime;
+  if (s != cs) {
+    setTime(s);
+    emitValueChanged();
+  }
+}
+
+void TimerEdit::setTimeRange(int minSeconds, int maxSeconds)
+{
+  if (minSeconds <= maxSeconds && (m_minTime != minSeconds || m_maxTime != maxSeconds)) {
+    int digits = 0;
+    if (maxSeconds >= 3600)
+      digits = floorf(logf(round(maxSeconds / 3600)) / logf(10.0f)) + 1;
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Uppercase the literal suffix. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3ALiteralSuffix)